src/cli/log.rs

```diff
@@ -1,94 +1,149 @@
+    ( $ ( $ arg : tt ) * ) => ( ::tracing::trace ! ( $ ( $ arg ) * )  )
 }
-macro_rules! err {
-    ( $ ( $ arg : tt ) * ) => ( $crate::cli::log::err_fmt ( format_args ! ( $ ( $ arg ) * ) ) )
+
+macro_rules! verbose {
+    ( $ ( $ arg : tt ) * ) => ( ::tracing::debug ! ( $ ( $ arg ) * )  )
 }
+
 macro_rules! info {
-    ( $ ( $ arg : tt ) * ) => ( $crate::cli::log::info_fmt ( format_args ! ( $ ( $ arg ) * ) ) )
+    ( $ ( $ arg : tt ) * ) => ( ::tracing::info ! ( $ ( $ arg ) * )  )
 }
 
-macro_rules! verbose {
-    ( $ ( $ arg : tt ) * ) => ( $crate::cli::log::verbose_fmt ( format_args ! ( $ ( $ arg ) * ) ) )
+macro_rules! warn {
+    ( $ ( $ arg : tt ) * ) => ( ::tracing::warn ! ( $ ( $ arg ) * )  )
 }
 
-macro_rules! debug {
-    ( $ ( $ arg : tt ) * ) => ( $crate::cli::log::debug_fmt ( format_args ! ( $ ( $ arg ) * ) ) )
+macro_rules! err {
+    ( $ ( $ arg : tt ) * ) => ( ::tracing::error ! ( $ ( $ arg ) * )  )
 }
 
-pub(crate) fn warn_fmt(args: fmt::Arguments<'_>) {
-    let mut t = process().stderr().terminal();
-    let _ = t.fg(terminalsource::Color::Yellow);
-    let _ = t.attr(terminalsource::Attr::Bold);
-    let _ = write!(t.lock(), "warning: ");
-    let _ = t.reset();
-    let _ = t.lock().write_fmt(args);
-    let _ = writeln!(t.lock());
-}
+pub fn tracing_subscriber(process: Process) -> impl tracing::Subscriber {
+    use tracing_subscriber::{layer::SubscriberExt, Registry};
 
-pub(crate) fn err_fmt(args: fmt::Arguments<'_>) {
-    let mut t = process().stderr().terminal();
-    let _ = t.fg(terminalsource::Color::Red);
-    let _ = t.attr(terminalsource::Attr::Bold);
-    let _ = write!(t.lock(), "error: ");
-    let _ = t.reset();
-    let _ = t.lock().write_fmt(args);
-    let _ = writeln!(t.lock());
+    #[cfg(feature = "otel")]
+    let telemetry = telemetry(process.clone());
+    let console_logger = console_logger(process);
+    #[cfg(feature = "otel")]
+    {
+        Registry::default().with(console_logger).with(telemetry)
+    }
+    #[cfg(not(feature = "otel"))]
+    {
+        Registry::default().with(console_logger)
+    }
 }
 
-pub(crate) fn info_fmt(args: fmt::Arguments<'_>) {
-    let mut t = process().stderr().terminal();
-    let _ = t.attr(terminalsource::Attr::Bold);
-    let _ = write!(t.lock(), "info: ");
-    let _ = t.reset();
-    let _ = t.lock().write_fmt(args);
-    let _ = writeln!(t.lock());
+/// A [`tracing::Subscriber`] [`Layer`][`tracing_subscriber::Layer`] that prints out the log
+/// lines to the current [`Process`]' `stderr`.
+///
+/// When the `RUST_LOG` environment variable is present, a standard [`tracing_subscriber`]
+/// formatter will be used according to the filtering directives set in its value.
+/// Otherwise, this logger will use [`EventFormatter`] to mimic "classic" Rustup `stderr` output.
+fn console_logger<S>(process: Process) -> impl Layer<S>
+where
+    S: Subscriber + for<'span> LookupSpan<'span>,
+{
+    let maybe_rust_log_directives = process.var_os("RUST_LOG").clone();
```

Comment:
Don't think we need the `clone()` here?

Also this feels like it's doign a bunch of manual reimplementation of things that tracing-subscriber offer for us already -- not sure what the goal is here?

Reply:
@djc The goal here is to mirror the original Rustup output when `RUST_LOG` is unset, see https://github.com/rust-lang/rustup/pull/3803#issuecomment-2092349333. On the other hand, the default formatter from `tracing_subsriber` is used once `RUST_LOG` is set.

Reply:
@djc FWIW, the current implementation gives:

```console
> RUSTUP_FORCE_ARG0=rustup rustup run stable cargo run -- default stable
     Running `target/debug/rustup-init default stable`
info: using existing install for 'stable-aarch64-apple-darwin'
info: default toolchain set to 'stable-aarch64-apple-darwin'

  stable-aarch64-apple-darwin unchanged - rustc 1.79.0 (129f3b996 2024-06-10)

info: note that the toolchain 'stable-aarch64-apple-darwin' is currently in use (overriden by environment variable RUSTUP_TOOLCHAIN)
```

However, once we set `RUST_LOG`:

```console
> RUST_LOG=trace RUSTUP_FORCE_ARG0=rustup rustup run stable cargo run -- default stable
     Running `target/debug/rustup-init default stable`
2024-06-15T02:22:11.801871Z  INFO rustup::cli::common: using existing install for 'stable-aarch64-apple-darwin'
2024-06-15T02:22:11.814343Z  INFO rustup::cli::common: default toolchain set to 'stable-aarch64-apple-darwin'

  stable-aarch64-apple-darwin unchanged - rustc 1.79.0 (129f3b996 2024-06-10)

2024-06-15T02:22:11.822315Z  INFO rustup::cli::rustup_mode: note that the toolchain 'stable-aarch64-apple-darwin' is currently in use (overriden by environment variable RUSTUP_TOOLCHAIN)
```

Reply:
> Don't think we need the `clone()` here?

@djc Fixed.

> Also this feels like it's doign a bunch of manual reimplementation of things that tracing-subscriber offer for us already -- not sure what the goal is here?

The goal here is to mirror the original Rustup output when `RUST_LOG` is not set (see https://github.com/rust-lang/rustup/pull/3803#issuecomment-2092349333 for the original plan):

```console
> RUSTUP_FORCE_ARG0=rustup rustup run stable cargo run -- default stable
     Running `target/debug/rustup-init default stable`
info: using existing install for 'stable-aarch64-apple-darwin'
info: default toolchain set to 'stable-aarch64-apple-darwin'

  stable-aarch64-apple-darwin unchanged - rustc 1.79.0 (129f3b996 2024-06-10)

info: note that the toolchain 'stable-aarch64-apple-darwin' is currently in use (overriden by environment variable RUSTUP_TOOLCHAIN)
```

This means that I won't have to change our integration tests too much (please note that there's no "refresh" mode for a whole bunch of handwritten tests, so keeping the old output is a very important aspect in the migration). That said, I'm totally open to changing the output format if we get to migrated to another test framework such as `snapbox` that supports refreshing in the future (like in https://github.com/rust-lang/cargo/issues/14039).

On the other hand, once `RUST_LOG` is set, **the default formatter from `tracing_subsriber` is used** instead:

```console
> RUST_LOG=trace RUSTUP_FORCE_ARG0=rustup rustup run stable cargo run -- default stable
     Running `target/debug/rustup-init default stable`
2024-06-15T02:22:11.801871Z  INFO rustup::cli::common: using existing install for 'stable-aarch64-apple-darwin'
2024-06-15T02:22:11.814343Z  INFO rustup::cli::common: default toolchain set to 'stable-aarch64-apple-darwin'

  stable-aarch64-apple-darwin unchanged - rustc 1.79.0 (129f3b996 2024-06-10)

2024-06-15T02:22:11.822315Z  INFO rustup::cli::rustup_mode: note that the toolchain 'stable-aarch64-apple-darwin' is currently in use (overriden by environment variable RUSTUP_TOOLCHAIN)
```

I'll leave the relevant doc changes to another PR, so that we can work on it in parallel with #3873 and #3871.